src/v/storage/disk_log_impl.cc

```diff
@@ -1308,6 +1308,22 @@ disk_log_impl::get_term_last_offset(model::term_id term) const {
     return std::nullopt;
 }
 
+std::optional<model::offset>
+disk_log_impl::previous_batch_boundary_offset(model::offset o) const {
+    if (unlikely(_segs.empty())) {
+        return std::nullopt;
+    }
+    auto it = _segs.lower_bound(o);
+    if (it == _segs.end()) {
+        return std::nullopt;
+    }
+    auto& idx = (*it)->index();
+    if (auto entry = idx.find_nearest(o)) {
```

Comment:
i think this will not work as expected. We use sparse index i.e. the nearest batch may not be the one closest to the offset but the one closest to the offset and indexed.

Reply:
If there is no way around this the expected behavior is that the actual data removed on disk will be slightly greater then if we did know the nearest segment boundary to the requested deletion offset, which is expected because a user could request to delete at any arbitrary index.

Opinions?

Reply:
>actual data removed on disk will be slightly greater

Wouldn't we want to avoid removing any data at offsets higher than requested, and if the offset were within a batch or in the middle of a segment then we'd do a logical truncation? What does kafka compatibility say about it?

Reply:
if we will stick to the current semantics of this method we must at least rename it to reflect the behavior

Reply:
@graphcareful i think this comment is still not addressed

Reply:
Thanks for reminding me, i think we agreed on a rename, would `nearest_known_batch_boundary_offset` make sense?

Reply:
Renamed to `index_lower_bound`